activerecord/lib/active_record/reflection.rb

```diff
@@ -466,7 +466,12 @@ def join_table
       end
 
       def foreign_key
-        @foreign_key ||= -(options[:foreign_key]&.to_s || derive_foreign_key)
+        @foreign_key ||= if options[:foreign_key] && options[:foreign_key].is_a?(Array)
+          # composite foreign keys support
+          options[:foreign_key].map(&:to_s).freeze
```

Comment:
The individual elements in this list are not being `freeze`d, whereas the single foreign key path freezes, the result. The danger is that someone could query foreign keys and modify an entry in-place,